proposals/ref-struct-interfaces.md

```diff
@@ -0,0 +1,195 @@
+Ref Struct Interfaces
+=====
+
+## Summary
+This proposal will expand the capabilities of `ref struct` such that they can implement interfaces and participate as generic type arguments.
+
+## Motivation
+The inability for `ref struct` to implement interfaces means they cannot participate in fairly fundamental abstraction techniques of .NET. A `Span<T>`, even though it has all the attributes of a sequential list cannot participate in methods that take `IReadOnlyList<T>`, `IEnumerable<T>`, etc ... Instead specific methods must be coded for `Span<T>` that have virtually the same implementation. Allowing `ref struct` to implement interfaces will allow operations to be abstracted over them as they are for other types. 
+
+## Detailed Design
+
+### ref struct interfaces
+<a name="ref-struct-interface"></a>
+The language will allow for `ref struct` types to implement interfaces. The syntax and rules are the same as for normal `struct` with a few exceptions to account for the limitations of `ref struct` types. 
+
+The ability to implement interfaces does not impact the existing limitations against boxing `ref struct` instances. That means even if a `ref struct` implements a particular interface,  it cannot be directly cast to it as that represents a boxing action.
+
+```csharp
+ref struct File : IDisposable
+{
+    private SafeHandle _handle;
+    public void Dispose()
+    {
+        _handle.Dispose();
+    }
+}
+
+File f = ...;
+// Error: cannot box `ref struct` type `File`
+IDisposable d = f;
+```
+
+The ability to implement interfaces is only useful when combined with the ability for `ref struct` to participate in generic arguments (as [laid out later](#ref-struct-generic)).
+
+To allow for interfaces to cover the full expressiveness of a `ref struct`, the language will allow `[UnscopedRef]` to appear on interface methods and properties. When a `ref struct` member implements an interface member with a `[UnscopedRef]` attribute, that `ref struct` member must also be decorated with `[UnscopedRef]`. The attribute is ignored when the `class` implements the interface. 
+
+Default interface methods pose a problem for `ref struct` as there are no protections against the default implementation boxing the `this` member. 
+
+```csharp
+interface I1
+{
+    void M()
+    {
+        // Error: both of these box if I1 is implemented by a ref struct
+        I1 local1 = this;
```

Comment:
The runtime will demand a `this` pointer as a boxed interface even without any of these explicit usages. Simply calling this method will break the runtime. So you're right that this needs a runtime exception.